### ISSUE TITLE
Translation to Spanish dont works in our customizations (Business Central OnPremise)

### ISSUE THREAD

**Issue by JavierFuentes** [Microsoft/AL#4232]
**Describe the bug**
`es-ES` translations (only for my customizations) are not working on my OnPremise Business Central server.

**To Reproduce**
I have added a new field to an standard Table and her Page (Customer Price Group in my case).

I have created a new Table and her new Page too.

In both cases, all names are with a custom prefix because I have `"mandatoryPrefix"` activated, and I use the **Caption property** to set the friendly visual name for all them...

Because I have `"features": ["TranslationFile"]` in my `app.json`, after every compilation, VSCode generates the translation file `\Translations\MyExtensionName.g.xlf` with "source-language" as `en-US` despite my version of Business Central is an `es-ES localization` (but I can't specify this in my project anywhere)...

I've copied this file with the name `MyExtensionName.es-ES.xlf` and I've edited to set "target-language" to `es-ES` and added with Spanish translations all <target></target> entries under every <source></source> ones.

When I deploy to my OnPremise server, I see al this Captions in English despite my active Language is `Spanish (Spain)` so my translations are not working...

**Expected behavior**
My users with `Spanish (Spain)` language active should see all my `es-ES` translations for my customizations

**Screenshots**
![image](https://user-images.githubusercontent.com/5801847/48709280-4ca9fb00-ec05-11e8-95ce-e52f7534c2f2.png)

**Versions:**
AL Language: 2.0.48254
Business Central: ES Dynamics NAV 13.0 (24630) OnPremise

**Comment by kalberes** [comment]
Have you tried adding a 
  "supportedLocales": [
    "es-ES"
  ]
to your app.json?

**Comment by JavierFuentes** [comment]
Hi @kalberes , 
yes I have this configuration in `app.json`

`"supportedLocales": ["en-US", "es-ES", "de-DE"],`

**Comment by thpeder** [comment]
Hi @JavierFuentes,
It is hard to follow your problem without a small sample project that reproduces the problem, but it sounds like you haven't set the target-language in the .xlf file.
Have a look at [Working with Translation Files](https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-work-with-translation-files#the-xliff-file) to learn more.

**Comment by StanislawStempin** [comment]
When deploying your extension do you see a message from the compiler that acknowledged your translation file being recognized as valid translation file? 
You seem to be doing all the right steps so I wonder if the issue is just with the content of your translation file. For example are all characters properly escaped.

**Comment by JavierFuentes** [comment]
Hi @StanislawStempin and @thpeder,
I attach here **a tiny Hello-World issue example with a Label translated to Spanish**... when I deploy this on my OnPremise Server the message is showed in English language and not in Spanish (the language by default when I log in).

[TranslationIssue.zip](https://github.com/Microsoft/AL/files/2612153/TranslationIssue.zip)

Within the ZIP I have included all the source code, the libraries of symbols that VSCode has downloaded from my server and even the extension downloaded in ZIP format

![image](https://user-images.githubusercontent.com/5801847/48966663-703ebe00-efd6-11e8-9e9b-809569498b8c.png)

I appreciate all the help you can give me to solve this issue.

**Comment by thpeder** [comment]
Hi @JavierFuentes,
I don't have spanish installed at the moment so I tried with danish, and when I set the region and the language to, in my case, Danish then I see the message otherwise it doesn't show up, because then the culture I'm using is something like da-US and not da-DK.
Can you try again with settings like this:
![image](https://user-images.githubusercontent.com/24276180/49152888-6c7fa400-f314-11e8-968b-5e846e0533bd.png)

**Comment by jaloag** [comment]
Hi Javier. 

If you espand the languages list you'llo see there are two "Spanish (Spain)". Try using the another language. 

Greetings.

**Comment by StanislawStempin** [comment]
Closing based on provided solution.

**Comment by JavierFuentes** [comment]
Hi @StanislawStempin,
I am sorry for not having answered before but it has been impossible for me to test after the last comments that have been added to this Ticket.

@jaloag, thanks... I have verified that indeed the `Español (España)` language appears twice in the list with the exact same description (which is totally confusing) and does so for the Language as well as for the Regions.

I have added to the original message the calls to the functions that return the numeric code of the active language:
```java
             Message (HelloMsg + '\\' +
                     'GlobalLanguage:' + Format (System.GlobalLanguage ()) + '\' +
                     'WindowsLanguage:' + Format (System.WindowsLanguage ()));
```

I'm sorry to say that **the result remains be wrong** with all possible combinations of Region and Language.

As you can see in this images, the message is showing in English and not in Spanish:

![image](https://user-images.githubusercontent.com/5801847/49589960-f362fb00-f96a-11e8-8e08-f1b1b7bc2813.png)
![image](https://user-images.githubusercontent.com/5801847/49589974-f9f17280-f96a-11e8-80e6-fcc4dbdc3974.png)
![image](https://user-images.githubusercontent.com/5801847/49589990-05449e00-f96b-11e8-834f-aea3a5e7f389.png)
![image](https://user-images.githubusercontent.com/5801847/49590002-0aa1e880-f96b-11e8-9f89-9ac145a0ced6.png)

NOTE.- In [this link](https://docs.microsoft.com/es-es/deployoffice/office2016/language-identifiers-and-optionstate-id-values-in-office-2016), I can see that 3082 Language ID is for es-ES Code.

### **With all these combinations, the standard labels for fields and tables are showing in Spanish all the time... even establishing the Español (Mexico) language**

@thpeder , I have tried to set `target-language="es-US"`, `target-language="es-ES"`, `target-language="es-es"` and `target-language="es"` but it doesn't worked for me...

Because `Español (España)` is the only one language apears twice, I think this could be a Bug of this version with Spanish custom translations...

Could you investigate it please?
Thanks for your help!

**Comment by thpeder** [comment]
Hi @JavierFuentes,
If there are no combination of Region and language that let's you see the translated texts then we have a bug.

**Comment by JavierFuentes** [comment]
Hello again,
I have some more information that I can give you about this issue ... 

**I have managed to make translations into Spanish work with several types of data,... `but they still fail with Label type.`**


**The keys** for the translations to work with the rest of the data types I've tested have been 2:
1. **_target-language_ must be equals to _"es"_**... "es-ES" doen't work!
2. keep the tags _**< source ></ source >**_ exactly with the same string it has in the _.g.xlf_ file before the _**< target ></ target >**_ tags in _.es.xlf_ file

I've updated my example and attach here again:
[TranslationIssue2.zip](https://github.com/Microsoft/AL/files/2682863/TranslationIssue2.zip)

### These are the modifications I've made:

![image](https://user-images.githubusercontent.com/5801847/50045255-cf509980-008f-11e9-9543-507014467517.png)

![image](https://user-images.githubusercontent.com/5801847/50045258-d4ade400-008f-11e9-94c9-5ccdfe47d568.png)

### An these are the results:
![image](https://user-images.githubusercontent.com/5801847/50045105-b8a94300-008d-11e9-8304-f08fe13a8ebe.png)

I hope these can help you to fix these issue with Label translations to Spanish.

**Comment by thpeder** [comment]
Hi @JavierFuentes,
The issue is because there exists mutliple spanish cultures, given from your GlobalLanguage with ID **1034** and your textConst with culture name **ESP**, the culture name you are looking for is **es-ES_tradnl** in your translation file, which the compiler incorrectly will report as invalid at the moment so I'm fixing that.

| Culture      | ID   | ThreeLetterWindowsName | EnglishName                         |
| -------------- | ------ | ------------------------ | ------------------------------------- |
| es           | 10   | ESP                    | Spanish                             |
| es-ES        | 3082 | SPA                    | Spanish (Spain, International Sort) |
| es-ES_tradnl | 1034 | ESP                    | Spanish (Spain, Traditional Sort)   |

**Comment by JavierFuentes** [comment]
Hi @thpeder,
I am glad to know that finally there is a bug and that it is not something that I am doing incorrectly or that it is badly installed on my OnPremise server.

Yes, 
I suppose that all we spanish-speakers are special people who need all those different encodings :-(

Culture Codes are relative to the format of decimals, currencies and dates, but here we talk about translations of text strings ...

I understand that the Culture code "es", when there were no more specific translations loaded in BC, should overwrite and work correctly with any other combination of Language and Region "es- *" that the user could have configured on their devices.

PD.- Thank you all for your efforts to solve this problem with the mother tongue of more than 442 million people in the world.

**Comment by marknitek** [comment]
Currently the compiler (2.1.79379) tells mit that de-CH  (2055) is not supported locale. Is this a related issue? Will this be fixed with the spring release?
It works when using de-DE.

**Comment by marknitek** [comment]
Spring release is here and the issue is still happening... We need to specify de-CH languages. 
When will this be possible?

![image](https://user-images.githubusercontent.com/2346803/55969144-9612e980-5c7d-11e9-9a6a-65941a3f285b.png)

**Comment by salgiza** [comment]
Las week I started a thread relating this problem in Yammer (I didn't know about this ticket).
https://www.yammer.com/dynamicsnavdev/#/threads/show?threadId=102064519135232

A new version of BC has been released, we still have two "Spanish (Spain)" options in the Language Selector, and our translations only apply correctly if you choose the correct one (the one after Australian English, as @josecarlosherrero found out).

If this is really related to Spanish "Traditional sort" and Spanish "International/sort", please, kill the "Traditional sort" with fire. Almost no-one knows what that means, it only applies to table collations when storing data, and it breaks search in BC if you choose it (but only in subtle ways that make it hard to notice; lots of fun).

**Comment by mouddai** [comment]
I have the same issu with fr-CA:
![image](https://user-images.githubusercontent.com/8635118/70163941-4c549480-168e-11ea-8426-f3ef0a14d08d.png)

**Comment by mouddai** [comment]
![image](https://user-images.githubusercontent.com/8635118/70163981-60989180-168e-11ea-95e3-571be5f77a9a.png)

**Comment by mouddai** [comment]
![image](https://user-images.githubusercontent.com/8635118/70164272-e0bef700-168e-11ea-92fe-76a749517b47.png)

**Comment by mouddai** [comment]
I also found that BC does not change language in a local development 
![image](https://user-images.githubusercontent.com/8635118/70164890-fb45a000-168f-11ea-8a38-7fb402b77789.png)

**Comment by marknitek** [comment]
**I finally managed to fixed the issue for me!** 
It turned out it was caused by a custom locale that we create for de-CH. This caused the compiler to not recognize the stirng `de-CH` in the app.json.  

I forgot about that locale since it was installed years ago...
![image](https://user-images.githubusercontent.com/2346803/71577488-a2cdbb00-2af4-11ea-9a2c-3b91fa661bfd.png)

Here you can see the hint that a "custom" locale was used (notice the *)
![image](https://user-images.githubusercontent.com/2346803/71577544-ea544700-2af4-11ea-8642-4296ada37fd5.png)
![image](https://user-images.githubusercontent.com/2346803/71577558-f17b5500-2af4-11ea-8de7-11bb1e80d925.png)

After uninstallation everything the * is gone
![image](https://user-images.githubusercontent.com/2346803/71577579-048e2500-2af5-11ea-8709-5078ecc0ee6e.png)

And the Compiler is happy
![image](https://user-images.githubusercontent.com/2346803/71577620-2a1b2e80-2af5-11ea-902b-4c060841c532.png)

**Comment by MarcusFabian** [comment]
I found another possibility (not mentioned above) why a Translation might not work!
You might have an xml error such as un-closed tag:
![image](https://user-images.githubusercontent.com/40990528/102616431-d3f53000-4137-11eb-81b8-e08691402406.png)

The Editor within VS does not do any validation of the XML-Fil

**Comment by maksym-ch** [comment]
Using es-ES_tradnl in translation file helped me [BC 20] + "supportedLocales": ["en-US","fr-BE","nl-BE","es-ES"] - I have BE localized version
